src/Lecture2.hs

```diff
@@ -261,7 +324,26 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval _ (Lit n) = Right n
+eval vars (Var x) = 
+  case lookup x vars  of 
+    Nothing -> Left (VariableNotFound x)
+    Just a -> Right a
+eval vars (Add left right) = sumEither (eval vars left)  (eval vars right)
+  where 
+    sumEither :: Either EvalError Int -> Either EvalError Int -> Either EvalError Int
+    sumEither first second =
+      case first of 
+        Left err1 -> 
+          case second of 
+                        Left err2 -> Left err2
+                        Right _ -> Left err1
+        Right a -> 
+          case second of 
+                        Left err -> Left err 
+                        Right b -> Right (a + b) 
```

Comment:
That's a great helper function! 👏🏻 
Once you learn about Monads, you'll be able to reduce this function to a single line ✂️